File: app/src/main/java/com/example/homeworkproject/repository/Repository.java
package com.example.homeworkproject.repository;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;

import com.example.homeworkproject.client.Api;
import com.example.homeworkproject.client.RetrofitClient;
import com.example.homeworkproject.model.ApiState;
import com.example.homeworkproject.model.Country;
import com.example.homeworkproject.model.Province;

import java.util.ArrayList;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class Repository {
    Api apiRequest = RetrofitClient.getMyApi();

    public MutableLiveData<ArrayList<Country>> getCountries(){
        final MutableLiveData<ArrayList<Country>> countryLiveData = new MutableLiveData<>();

        Call<ArrayList<Country>> call = apiRequest.getCountry();

        call.enqueue(new Callback<ArrayList<Country>>() {
            @Override
            public void onResponse(@NonNull Call<ArrayList<Country>> call, @NonNull Response<ArrayList<Country>> response) {

                if (response.isSuccessful() && response.body() !=null){

                    countryLiveData.setValue(response.body());
                } }

            @Override
            public void onFailure(@NonNull Call<ArrayList<Country>> call, @NonNull Throwable t) {
                ApiState.error();
            }
        });

        return countryLiveData;
    }

    public MutableLiveData<ArrayList<Province>> getProvince(String value) {

        final MutableLiveData<ArrayList<Province>> provinceLiveData = new MutableLiveData<>();
        Call<ArrayList<Province>> call = apiRequest.getProvince(value);
        call.enqueue(new Callback<ArrayList<Province>>() {
            @Override
            public void onResponse(@NonNull Call<ArrayList<Province>> call, @NonNull Response<ArrayList<Province>> response) {
                if (response.isSuccessful()){
                    provinceLiveData.setValue(response.body());
                }
            }

            @Override
            public void onFailure(@NonNull Call<ArrayList<Province>> call, @NonNull Throwable t) {
                ApiState.error();
            }
        });
        return provinceLiveData;
    }
}

File: app/src/main/java/com/example/homeworkproject/repository/Repository.kt
package com.example.homeworkproject.repository

import androidx.lifecycle.MutableLiveData
import com.example.homeworkproject.client.RetrofitClient.myApi
import com.example.homeworkproject.model.Country
import com.example.homeworkproject.model.Province
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class Repository {

    private var apiRequest = myApi

    fun getCountries(): MutableLiveData<ArrayList<Country>> {
        val countryLiveData= MutableLiveData<ArrayList<Country>>()
        val call: Call<ArrayList<Country>> = apiRequest.getCountry()
        call.enqueue(object: Callback<ArrayList<Country>>{
            override fun onResponse(
                call: Call<ArrayList<Country>>,
                response: Response<ArrayList<Country>>
            ) {
                if (response.isSuccessful && response.body() != null) {
                    countryLiveData.value = response.body()
                }
            }

            override fun onFailure(call: Call<ArrayList<Country>>, t: Throwable) {
                countryLiveData.value = null
            }
        })
        return countryLiveData
    }

    fun getProvinces(countryId: String?): MutableLiveData<ArrayList<Province>> {
        val provinceLiveData= MutableLiveData<ArrayList<Province>>()
        val call: Call<ArrayList<Province>> = apiRequest.getProvince(countryId)
        call.enqueue(object: Callback<ArrayList<Province>>{
            override fun onResponse(
                call: Call<ArrayList<Province>>,
                response: Response<ArrayList<Province>>
            ) {
                if (response.isSuccessful && response.body() != null) {
                    provinceLiveData.value = response.body()
                }
            }

            override fun onFailure(call: Call<ArrayList<Province>>, t: Throwable) {
                provinceLiveData.value = null
            }
        })
        return provinceLiveData
    }

}
